paper-server/src/main/java/io/papermc/paper/connection/HorriblePlayerLoginEventHack.java

```diff
@@ -0,0 +1,90 @@
+package io.papermc.paper.connection;
+
+import com.mojang.authlib.GameProfile;
+import com.mojang.logging.LogUtils;
+import io.papermc.paper.adventure.PaperAdventure;
+import net.minecraft.network.Connection;
+import net.minecraft.network.chat.Component;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ClientInformation;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.players.PlayerList;
+import net.minecraft.world.level.Level;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.event.player.PlayerLoginEvent;
+import org.bukkit.plugin.RegisteredListener;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+// TODO: DO we even need this? Who mutates the player here? it does nothing anyways.
+// TODO: Make sure player settings are migrated too
+public class HorriblePlayerLoginEventHack {
+
+    private static final Logger LOGGER = LogUtils.getClassLogger();
+    private static boolean nagged = false;
+
+    @Nullable
+    public static Component execute(final Connection connection, MinecraftServer minecraftServer, GameProfile profile, @NotNull PlayerList.LoginResult result) {
+        if (PlayerLoginEvent.getHandlerList().getRegisteredListeners().length == 0) {
+            return result.message();
+        }
+
+        if (!nagged) {
+            List<String> plugins = new ArrayList<>();
+            for (final RegisteredListener listener : PlayerLoginEvent.getHandlerList().getRegisteredListeners()) {
+                plugins.add(listener.getPlugin().getName());
+            }
+
+            LOGGER.warn("""
+                
+                ============================================================
+                WARNING: Legacy PlayerLoginEvent usage detected!
+                
+                This event forces an alternative player loading path that is
+                deprecated and will be REMOVED in a future release.
+                
+                This also does not properly represent the fact that this validation logic is ran
+                twice in vanilla.
+                
+                Please notify the following plugin developers: %s
+
+                Continued usage may cause instability in future builds.
```

Comment:
This should be shortened/reworded a bit, maybe use a link to the javadocs instead of that "See ..." line:

```suggestion
                WARNING: Legacy PlayerLoginEvent usage detected!
                
                This event forces an alternative player loading path that is
                deprecated and will be removed in a future release.
                See the PlayerLoginEvent javadocs for more information.
                
                Please notify the following plugin developers: %s
```